### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by ammarali32**

### CODE REVIEW

File: src/main/java/impl/image/ConvolutionProviderFactory.java [hunk 1]
```diff
@@ -2,8 +2,41 @@
 
 import api.image.ConvolutionProvider;
 
+import java.awt.*;
+import java.lang.Math.*;
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderFactory.java]: Redundant import

File: src/main/java/impl/image/ConvolutionProviderFactory.java [hunk 2]
```diff
@@ -2,8 +2,41 @@
...
 public class ConvolutionProviderFactory {
-    public static ConvolutionProvider getInstance() {
-        return null;
+    private static ConvolutionProviderFactory instance;
+    public static ConvolutionProviderFactory getInstance()
+    {
+        if (instance == null)instance = new ConvolutionProviderFactory();
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderFactory.java]: You should create the new class which implements `ConvolutionProvider` interface and return it's object here.  
Method signature should stay `public static ConvolutionProvider getInstance()`

Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderFactory.java]: It is important to use `{` `}` for all expression types. 
https://www.oracle.com/technetwork/java/javase/documentation/codeconventions-142311.html#431

File: src/main/java/impl/image/ConvolutionProviderFactory.java [hunk 3]
```diff
@@ -2,8 +2,41 @@
...
+        return instance;
     }
+    public Color[][] apply(Color[][] image, double[][] kernel){
+        int n=image.length,m=image[0].length,kn =kernel.length,km=kernel[0].length;
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderFactory.java]: You should initialize one variable per line.
https://www.oracle.com/technetwork/java/javase/documentation/codeconventions-141270.html#2992

Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderFactory.java]: Needs more spaces, please, follow https://www.oracle.com/technetwork/java/codeconvtoc-136057.html

File: src/main/java/impl/image/ConvolutionProviderFactory.java [hunk 4]
```diff
@@ -2,8 +2,41 @@
...
+        Color[][] img = new Color[n][m];
+        for(int i=0;i<n;i++){
+            for(int j=0;j<m;j++){
+                int blue=0,red=0,green=0;
+                int id1=0,id2 =0;
+                for (int it1=-(kn/2);it1<=(kn/2);it1++){
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderFactory.java]: Not enough spaces

Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderFactory.java]: You use expression `kn/2` , `km/2` expression more than one time, so extract it to a variable

File: src/main/java/impl/image/ConvolutionProviderFactory.java [hunk 5]
```diff
@@ -2,8 +2,41 @@
...
+                    id2=0;
+                    for(int it2=-(km/2);it2<=(km/2);it2++) {
+                        if (j + it2 < 0 || i + it1 < 0 || j + it2 >= m || i + it1 >= n) {id2+=1;continue;}
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderFactory.java]: new line needed

File: src/main/java/impl/image/ConvolutionProviderFactory.java [hunk 6]
```diff
@@ -2,8 +2,41 @@
...
+                        blue += (int)(kernel[id1][id2] * image[i + it1][j + it2].getBlue());
+                        green += (int)(kernel[id1][id2] * image[i + it1][j + it2].getGreen());
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderFactory.java]: Extract variables

File: src/main/java/impl/image/ConvolutionProviderFactory.java [hunk 7]
```diff
@@ -2,8 +2,41 @@
...
+                        red += (int)(kernel[id1][id2] * image[i + it1][j + it2].getRed());
+                        id2 += 1;
+                    }
+                    id1+=1;
+                }
+                img[i][j]=new Color((int)(red),(int)green,(int)blue,255);;
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderFactory.java]: Redundant `int` casting

File: src/main/java/impl/image/ConvolutionProviderFactory.java [hunk 8]
```diff
@@ -2,8 +2,41 @@
...
+            }
+        }
+
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderFactory.java]: Redundant empty spaces

File: src/main/java/impl/image/ConvolutionProviderFactory.java (hunk 4 above)
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderFactory.java]: Please, use friendly names for your variables. It is really hard to understand what do these variables mean

File: src/main/java/impl/image/ImageConverterFactory.java [hunk 9]
```diff
@@ -2,8 +2,51 @@
...
 import api.image.ImageConverter;
 
+import java.awt.*;
+
 public class ImageConverterFactory {
-    public static ImageConverter getInstance() {
-        return null;
+    private static ImageConverterFactory instance;
+    public static ImageConverterFactory getInstance() {
+        if(instance==null)instance =new ImageConverterFactory();
```
Comment (Walingar) [inline src/main/java/impl/image/ImageConverterFactory.java]: Same as with `ConvolutionProviderFactory`

File: src/main/java/impl/image/ConvolutionProviderFactory.java (hunk 7 above)
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderFactory.java]: Redundant `;`

File: src/main/java/impl/image/ImageConverterFactory.java [hunk 10]
```diff
@@ -2,8 +2,51 @@
...
+        return instance;
+    }
```
Comment (Walingar) [inline src/main/java/impl/image/ImageConverterFactory.java]: Need empty lines between methods, fields etc.

File: src/main/java/impl/image/ImageConverterFactory.java [hunk 11]
```diff
@@ -2,8 +2,51 @@
...
+    public Color[][] convertToColor(int [][] image) {
+        Color[][] img = new Color[image.length][image[0].length];
+        for (int i = 0; i < image.length; i++) {
+            for (int j = 0; j < image[0].length; j++) {
```
Comment (Walingar) [inline src/main/java/impl/image/ImageConverterFactory.java]: Extract variable

File: src/main/java/impl/image/ImageConverterFactory.java [hunk 12]
```diff
@@ -2,8 +2,51 @@
...
+                int blue = 0, green = 0, red = 0, alpha = 255;
+                for (int k = 0; k <= 23; k++) {
+                    if (k <= 7) blue += ((1 << k) & (image[i][j]));
+                    else if (k <= 15) green += (((1 << (k)) & (image[i][j]))>>8);
+                    else if (k <= 23) red +=(((1 << (k) & (image[i][j])))>>16);
+                }
+                img[i][j] = new Color(red, green, blue, alpha);
```
Comment (Walingar) [inline src/main/java/impl/image/ImageConverterFactory.java]: Please, read `Color` documentation and try to find prettier solution

File: src/main/java/impl/image/ImageConverterFactory.java [hunk 13]
```diff
@@ -2,8 +2,51 @@
...
+            }
+        }
+        return img;
+    }
+    public int[][] convertToRgb(Color[][]image)
+    {int[][] img = new int[image.length][image[0].length];
+        for(int i=0;i<image.length;i++){
+            for(int j=0;j<image[0].length;j++){
+                int blue=image[i][j].getBlue();
```
Comment (Walingar) [inline src/main/java/impl/image/ImageConverterFactory.java]: Also, read `Color` documentation. Solution can be better

File: src/main/java/impl/image/ImageConverterFactory.java [hunk 14]
```diff
@@ -2,8 +2,51 @@
...
+    }
+    public Color[][] convertToColor(int [][] image) {
+        Color[][] img = new Color[image.length][image[0].length];
+        for (int i = 0; i < image.length; i++) {
+            for (int j = 0; j < image[0].length; j++) {
+                int blue = 0, green = 0, red = 0, alpha = 255;
+                for (int k = 0; k <= 23; k++) {
+                    if (k <= 7) blue += ((1 << k) & (image[i][j]));
+                    else if (k <= 15) green += (((1 << (k)) & (image[i][j]))>>8);
+                    else if (k <= 23) red +=(((1 << (k) & (image[i][j])))>>16);
```
Comment (Walingar) [inline src/main/java/impl/image/ImageConverterFactory.java]: `k <= 23` is always true

File: src/main/java/api/weather/DayTemperatureInfo.java [hunk 15]
```diff
@@ -0,0 +1,11 @@
+package api.weather;
+
+import java.time.Month;
+
+public interface DayTemperatureInfo {
```
Comment (Walingar) [inline src/main/java/api/weather/DayTemperatureInfo.java]: Please, remove the 3rd task from this review. 
See https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History
You need to remove the last commit from your `master` branch

File: src/main/java/api/image/ImageConverter.java [hunk 16]
```diff
@@ -1,9 +1,6 @@
 package api.image;
-
 import java.awt.*;
-
```
Comment (Walingar) [inline src/main/java/api/image/ImageConverter.java]: These spaces are needed :)

Comment (Walingar) [reply src/main/java/api/image/ImageConverter.java]: Please, use Reformat Code action if you are using IntelliJ IDEA

File: src/main/java/impl/image/ConvolutionProviderImp.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
+package impl.image;
+
+import api.image.ConvolutionProvider;
+
+import java.awt.*;
+
+public class ConvolutionProviderImp implements ConvolutionProvider {
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImp.java]: Impl may be?

File: src/main/java/impl/image/ConvolutionProviderImp.java [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+        Color[][] img = new Color[n][m];
+        for (int i = 0; i < n; i++) {
+            for (int j = 0; j < m; j++) {
+                int blue = 0;
+                int red = 0;
+                int green = 0;
+                int index1 = 0;
+                int index2 = 0;
+                for (int iterator1 = -HalfKernelN; iterator1 <= HalfKernelN; iterator1++) {
+                    index2 = 0;
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImp.java]: What is this index? Why do you need it? It is not clear

**Comment by Walingar** [comment]
Please read this topic about commit messages. https://chris.beams.io/posts/git-commit/

File: src/main/java/api/image/ImageConverter.java (hunk 16 above)
Comment (ammarali32) [reply src/main/java/api/image/ImageConverter.java]: +

File: src/main/java/api/image/ImageConverter.java [hunk 19]
```diff
@@ -1,9 +1,6 @@
...
 public interface ImageConverter {
     Color[][] convertToColor(int[][] image);
-
```
Comment (Walingar) [inline src/main/java/api/image/ImageConverter.java]: Do you forget to add lines here?)

File: src/main/java/impl/image/ConvolutionProviderImp.java [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+                        Color imgVal = image[i + iterator1][j + iterator2];
+                        blue += (int) (kerenlVal * imgVal.getBlue());
+                        green += (int) (kerenlVal * imgVal.getGreen());
+                        red += (int) (kerenlVal * imgVal.getRed());
+                        index2 += 1;
+                    }
+                    index1 += 1;
+                }
+                img[i][j] = new Color(red, green, blue, 255);
+                ;
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImp.java]: Extra semicolon

File: src/main/java/impl/image/ConvolutionProviderImp.java [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+                int index1 = 0;
+                int index2 = 0;
+                for (int iterator1 = -HalfKernelN; iterator1 <= HalfKernelN; iterator1++) {
+                    index2 = 0;
+                    for (int iterator2 = -HalfKernelM; iterator2 <= HalfKernelM; iterator2++) {
+                        if (j + iterator2 < 0 || i + iterator1 < 0 || j + iterator2 >= m || i + iterator1 >= n) {
+                            index2 += 1;
+                            continue;
+                        }
+                        double kerenlVal = kernel[index1][index2];
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImp.java]: typo

File: src/main/java/impl/image/ConvolutionProviderImp.java [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+        int kernelM = kernel[0].length;
+        int HalfKernelN = kernelN / 2;
+        int HalfKernelM = kernelM / 2;
+        Color[][] img = new Color[n][m];
+        for (int i = 0; i < n; i++) {
+            for (int j = 0; j < m; j++) {
+                int blue = 0;
+                int red = 0;
+                int green = 0;
+                int index1 = 0;
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImp.java]: It is still not clear for me: why do you need these indices and what do they mean? It is better to use more clear wording

File: src/main/java/api/image/ImageConverter.java [hunk 23]
```diff
@@ -1,9 +1,6 @@
 package api.image;
```
Comment (Walingar) [inline src/main/java/api/image/ImageConverter.java]: And it is better not to modify api if it is not needed

File: src/main/java/impl/image/ConvolutionProviderImp.java (hunk 17 above)
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImp.java]: It is usually called ...Impl

**Pull request merged**